### DISCUSSION TITLE
Porting to Android?

### DISCUSSION THREAD

**Discussion by samuk** [markqvist/Reticulum#20] [Ideas]
This is probably silly, but now Rnode is running on generic ESP32, could Reticulum be ported to Android as an app? Would a USB or even Bluetooth connection to the ESP32 then work?

**Comment by stevewa2066** [comment]
> This is probably silly, but now Rnode is running on generic ESP32, could Reticulum be ported to Android as an app? Would a USB or even Bluetooth connection to the ESP32 then work?
> 
> 
> 
> 

Where is the code for the ESP32?

I have the parts for doing a Bluetooth on the original Rnode. Just haven't had the time.

**Comment by 4c3e** [comment]
@stevewa2066 I was able to install rnode software onto my ESP32 board using https://github.com/markqvist/rnodeconfigutil

Edit: the ESP32 code itself is in this repo: https://github.com/markqvist/RNode_Firmware

@samuk Nomadnet runs well on termux on Android ;)
For a native Android app, I ran into Kivy GUI library which would allow for an Android GUI app that could use RNS. The alternative could be to write a Kotlin Reticulum port which is something I'm increasingly interested in doing.

**Comment by samuk** [comment]
Sorry I'm a bit of a noob here: So Nomadnet contains the entire Reticulum stack? All you need is an Android device and an ESP32 device to join the mesh & communicate?

Do you need root? Or could this be packaged up into a Play friendly app?

**Comment by 4c3e** [comment]
@samuk Nomadnet runs its own Reticulum instance if you aren't already running the rnsd utility in the background already. On android you can run Nomadnet using Termux which is a terminal emulator for Android. There are install steps in the Reticulum manual. I've only tried using the TCP transport from Android(TCP Client) -> Computer(TCP Server). LoRa should:tm: work by plugging in a LoRa board to your Android phone using a USB cabled, but I have not tested it because I don't have a cable that fits (yet).

**Comment by markqvist** [comment]
_I wrote this comment this morning, and apparently was scatter-brained enough to leave my computer with the tab open without submitting it. Luckily for me it was still in memory (the computers, not mine), so here it is, some 14 hours later_

I already started implementing Android support for Reticulum a while ago, and the entire stack is actually already functional on Android :) Due to differences in hardware access and drivers in Android, what is currently missing is a way to connect directly to devices such as RNodes and modems. I am working on this problem _right now_. It is not entirely trivial, so it needs to be done right to not be a pain later, but is is completely doable, and it will be ready very soon.

@samuk Yes and no. Nomadnet is easy to install on Android in Termux. See this part of the manual: https://markqvist.github.io/Reticulum/manual/gettingstartedfast.html#reticulum-on-android. After you have run "pip3 install rns", just run "pip3 install nomadnet", and you are ready to go. Currently you will need another device on your local network that hosts the LoRa interface. By default, the AutoInterface will be enabled out of the box on Android. If you add an RNode to another system on your network that also has the AutoInterface enabled, and enable Transport in that systems config, your Android device can also communicate over the LoRa interface. No root is needed. That was one of the original intentions of implementing Reticulum as a purely-userland stack, without any dependencies on kernel calls.

@4c3e Yes, Kivy works well for these purposes, and the associated py4a tools allows building Reticulum directly into Android APKs. It works very well, and **I already have working prototype APKs** with the Reticulum stack included, that so far runs well on all the Android devices I have tested them on.

My current priorities in these regards are: First, implementing the necessary hardware interfacing for Reticulum to connect directly to hardware devices on Android. Second, implement an Android app with messaging features (that is interoperable with the rest of the ecosystem of course).

**Comment by samuk** [comment]
"Second, implement an Android app with messaging features (that is interoperable with the rest of the ecosystem of course)."

I wonder if it's worth forking https://briarproject.org/ they already have a nice UI worked out and 0.5million users. They might even accept a pull request upstream..

**Comment by 4c3e** [comment]
@samuk @markqvist I am an active Briar contributor 🙂 If there is any interest here, I'd be happy to make introductions and loop you all into the relevant channels. 
Briar project is currently doing their own independent mesh research, but I believe Reticulum could be a very useful transport to implement into Briar down the line.

**Reply by samuk** [reply]
@4c3e I posted this on the Briar Gitlab: https://code.briarproject.org/briar/briar/-/issues/1742#note_72698 you might have a more useful perspective on how Briar could interoperate with Reticulum

**Reply by markqvist** [reply]
I tried signing up for the Briar code repo a few days ago to share some thoughts on that thread, but so far my account is still pending approval. Maybe I have been categorised as a bot :S

**Reply by 4c3e** [reply]
@samuk thanks for dropping some info on that ticket! 

I definitely have spent a lot of time thinking about how Briar and Reticulum fit in together, and can share some thoughts :smile: 
For some context, Briar is built upon a p2p data sync protocol called Bramble (https://code.briarproject.org/briar/briar-spec/blob/master/protocols/BSP.md). Like Reticulum, the Bramble sync protocol can work on any arbitrary transport layer that supports one or two directional byte-streams. So it is definitely possible to implement Bramble-over-Reticulum. 

Another interesting development is Briar's mailbox project (https://code.briarproject.org/briar/briar-mailbox). This is going to start out as a program to outsource the background-polling Briar does on your device to a separate device, allowing users to set up a completely optional server to send and receive messages. Currently the mailbox app only works over tor, but it'd be interesting to see how something like LXMF would pair with this idea because there is a lot of overlap.

 @markqvist apologies for the wait! your account should be accepted soon.

**Comment by markqvist** [comment]
I have converted this thread to a discussion, since it is not an issue per se. At this point I think a discussion thread is more useful for updates on the progress and discussion around the subject.

**Comment by markqvist** [comment]
A small update on progress regarding this. The complete Reticulum stack, along with LXMF is now completely functional on Android, supporting versions from 8.1 through 12. I am writing a basic LXMF client for Android, that allows communication over Reticulum networks.

This photo shows the app running on a physical Android device (a Gemini PDA), receiving messages from a computer running the Nomadnet client.

![743b5f2d-0f3c-44e6-9854-9751906332b7](https://user-images.githubusercontent.com/1438012/158208818-a7d905c3-9185-41ac-b2d1-858070e43135.jpg)

**Reply by samuk** [reply]
What is LXMF please?

**Reply by markqvist** [reply]
It is an extremely bandwidth efficient, general-purpose message/data/command transfer protocol with delay- and disruption-tolerance, distributed and end-to-end encrypted on-network message-store, and minimal setup and implementation cost.

https://github.com/markqvist/lxmf

Nomadnet is an LXMF client, as is the above pictured Android app.

You can think of LXMF as an efficient and general-purpose solution for implementing a wide range of application "types" such as instant messaging, email, discussion boards, BBS'es and many other things.

**Comment by markqvist** [comment]
The first, and admittedly very early, alpha release of the Sideband application for Android is now out. You can download it and give it a spin over here: https://unsigned.io/sideband

**Reply by markqvist** [reply]
Currently, you will need a Reticulum gateway on your local WiFi network to access LoRa, packet radio or other interfaces from the Android app. I will add support for connecting radios and modems to the Android app soon, but for now you will need a transport node on the network your phone is connected to.

**Reply by samuk** [reply]
Looks good. I Don't have a Reticulum device, but the UI looks nice.

**Reply by OtisByron** [reply]
Very excited about this! Another piece of the puzzle coming together.

**Comment by Eninspace** [comment]
Hi there. Is there any way to connect my T-beam, via bluetooth or wifi, I tried it via sideband, but it didn't work for me. Then I installed Reticulum stack on termux. I tried this method https://www.reddit.com/r/androiddev/comments/ceadc6/comment/eudhl2k, but it also didn’t work for me, when I try to connect (rnodeconf --info) to the virtual serial port, an access error occurs. Is it possible to share wifi with esp32, would this be the simplest and most effective option, I think?

I used:
Tbeam 1.1 (Firmware version: 1.52)
rnodeconf 2.0.0
Android 13

**Reply by markqvist** [reply]
Well, it very much *should* work over Bluetoot with the T-Beam. I actually have a couple of T-Beams myself, that I use over Bluetooth.

If you can tell us exactly what goes wrong when trying it in Sideband, we should be able to get it sorted for you. But it's probably better to create a separate post here specifically for your issue. Make sure you are using the latest Sideband version, `v0.4.0`, since there was a Bluetooth bug in Sideband `v0.3.0` on some Android versions.

Under Termux itself, it will *not* work, unfortuately, since the Termux environment does not have access to USB or Bluetooth. It seems that this is unlikely to change from the Termux side, so to use Reticulum over Bluetooth and USB-based hardware, it needs to happen from one of the native Android apps that include Reticulum (since these can ask for permission to the USB and Bluetooth hardware layers).

In the future, there will be a standalone Reticulum app for Android that will provide a system-wide interface with access to USB and Bluetooth-based hardware, but it is not released yet.

The version 2.x line of the RNode firmware will also allow connecting to RNode devices over WiFi, in addition to USB and Bluetooth.

**Reply by Eninspace** [reply]
Ok, I will reproduce it in few days, and if will are problems I will create issue and try collect logs
